src/jit/morph.cpp

```diff
@@ -17987,569 +17971,763 @@ GenTree* Compiler::fgMorphImplicitByRefArgs(GenTree* tree, bool isAddr)
+    //
+    // To avoid this issue, the semantics of GT_FIELD (and for simplicity's sake any other
+    // indirection) nodes slightly deviates from the IR semantics - an indirection does not
+    // actually produce an unknown value but a location value, if the indirection address
+    // operand is an address value.
+    //
+    // The actual indirection is performed when the indirection's user node is processed:
+    //   - A GT_ADDR user turns the location value produced by the indirection back
+    //     into an address value.
+    //   - Any other user node performs the indirection and produces an unknown value.
+    //
+    class Value
     {
-        case GT_IND:
-            if (axc != AXC_Addr)
-            {
-                axcStack->Push(AXC_Ind);
-            }
-            else
+        GenTree* m_node;
+        unsigned m_lclNum;
+        unsigned m_offset;
+        bool     m_address;
+#ifdef DEBUG
+        bool m_consumed;
+#endif
+
+    public:
+        // TODO-Cleanup: This is only needed because ArrayStack initializes its storage incorrectly.
+        Value()
+            : m_node(nullptr)
+            , m_lclNum(BAD_VAR_NUM)
+            , m_offset(0)
+            , m_address(false)
+#ifdef DEBUG
+            , m_consumed(false)
+#endif
+        {
+        }
+
+        // Produce an unknown value
+        Value(GenTree* node)
+            : m_node(node)
+            , m_lclNum(BAD_VAR_NUM)
+            , m_offset(0)
+            , m_address(false)
+#ifdef DEBUG
+            , m_consumed(false)
+#endif
+        {
+        }
+
+        // The node that produced this value
+        GenTree* Node() const
+        {
+            return m_node;
+        }
+
+        // Does this value represent a location?
+        bool IsLocation() const
+        {
+            return (m_lclNum != BAD_VAR_NUM) && !m_address;
+        }
+
+        // Does this value represent the address of a location?
+        bool IsAddress() const
+        {
+            assert((m_lclNum != BAD_VAR_NUM) || !m_address);
+
+            return m_address;
+        }
+
+        // The location's variable number
+        unsigned LclNum() const
+        {
+            assert(IsLocation() || IsAddress());
+
+            return m_lclNum;
+        }
+
+        // The location's byte offset
+        unsigned Offset() const
+        {
+            assert(IsLocation() || IsAddress());
+
+            return m_offset;
+        }
+
+        // Produce a location value:
+        //   - (lclnum, offset) => LOCATION(lclNum, offset)
+        void Location(unsigned lclNum, unsigned offset = 0)
+        {
+            assert(!IsLocation() && !IsAddress());
+
+            m_lclNum = lclNum;
+            m_offset = offset;
+        }
+
+        // Produce an address value from a location value
+        //   - LOCATION(lclNum, offset) => ADDRESS(lclNum, offset)
+        //   - ADDRESS(...) => invalid
+        //   - UNKNOWN => UNKNOWN
```

Comment:
Hmm, are you referring to ADDRESS(...)? That's really ADDRESS(lclNum, offset), I just omitted them because they're no relevant. I added them back for clarity and explained that this should not happen because something like GT_ADDR(GT_ADDR(...)) is not valid.